File: src/android/GameAnalyticsPlugin.java
package com.gameanalytics.plugin.GameAnalyticsPlugin;

import com.gameanalytics.sdk.*;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**
 * This class echoes a string called from JavaScript.
 */
public class GameAnalyticsPlugin extends CordovaPlugin {
  // Standard Debugging Variables
  private static final String LCAT = "GameAnalyticsPlugin";
  /**
   * <p>
   * cordova.exec() method reference.
   * All exec() calls goes to this part.
   * </p>
   *
   * @param action          A {@link String} value method to be executed.
   * @param args            A {@link JSONArray} value parameters passed along with the action param.
   * @param callbackContext A {@link CallbackContext} function passed for executing callbacks.
   */
  @Override
  public boolean execute(String action, JSONArray args, final CallbackContext callback) throws JSONException {
    if (action.equals("initialize")) {
        final String gameKey = args.getString(0);
        final String gameSecret = args.getString(1);
        
        cordova.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                GameAnalytics.initializeWithGameKey(cordova.getActivity(), gameKey, gameSecret);
            }
        });
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureBuild")) {
        String build = args.getString(0);
        GameAnalytics.configureBuild(build);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureGameEngineVersion")) {
        String version = args.getString(0);
        GameAnalytics.configureGameEngineVersion(version);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureSdkGameEngineVersion")) {
        String version = args.getString(0);
        GameAnalytics.configureSdkGameEngineVersion(version);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureUserId")) {
        String id = args.getString(0);
        GameAnalytics.configureUserId(id);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addBusinessEvent")) {
        String currency = args.getString(0);
        int amount = args.getInt(1);
        String itemType = args.getString(2);
        String itemId = args.getString(3);
        String cartType = args.getString(4);
        GameAnalytics.addBusinessEventWithCurrency(currency, amount, itemType, itemId, cartType);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addBusinessEventWithReceipt")) {
        String currency = args.getString(0);
        int amount = args.getInt(1);
        String itemType = args.getString(2);
        String itemId = args.getString(3);
        String cartType = args.getString(4);
        String receipt = args.getString(5);
        String store = args.getString(6);
        String signature = args.getString(7);
        GameAnalytics.addBusinessEventWithCurrency(currency, amount, itemType, itemId, cartType, receipt, store, signature);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addResourceEvent")) {
        int flowType = args.getInt(0);
        String currency = args.getString(1);
        int amount = args.getInt(2);
        String itemType = args.getString(3);
        String itemId = args.getString(4);
        // 0 - source (add/buy), 1 - sink (subtract)
        GameAnalytics.addResourceEventWithFlowType(flowType, currency, amount, itemType, itemId);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addProgressionEvent")) {
        int status = args.getInt(0);
        String progression1 = args.getString(1);
        String progression2 = args.getString(2);
        String progression3 = args.getString(3);
        // 1 - Start , 2 - Complete , 3 - Fail
        GameAnalytics.addProgressionEventWithProgressionStatus(status, progression1, progression2, progression3);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addProgressionEventWithScore")) {
        int status = args.getInt(0);
        String progression1 = args.getString(1);
        String progression2 = args.getString(2);
        String progression3 = args.getString(3);
        int score = args.getInt(4);
        // 1 - Start , 2 - Complete , 3 - Fail
        GameAnalytics.addProgressionEventWithProgressionStatus(status, progression1, progression2, progression3, score);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addDesignEvent")) {
        String eventId = args.getString(0);
        GameAnalytics.addDesignEventWithEventId(eventId);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addDesignEventWithValue")) {
        String eventId = args.getString(0);
        int value = args.getInt(1);
        GameAnalytics.addDesignEventWithEventId(eventId, value);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("addErrorEvent")) {
        int severity = args.getInt(0);
        String message = args.getString(1);
        GameAnalytics.addErrorEventWithSeverity(severity, message);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setEnabledInfoLog")) {
        //GameAnalytics.setEnabledInfoLog(true);
        boolean flag = args.getBoolean(0);
        GameAnalytics.setEnabledInfoLog(flag);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setEnabledVerboseLog")) {
        //GameAnalytics.setEnabledVerboseLog(true);
        boolean flag = args.getBoolean(0);
        GameAnalytics.setEnabledVerboseLog(flag);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setEnabledManualSessionHandling")) {
        //GameAnalytics.setEnabledManualSessionHandling(true);
        boolean flag = args.getBoolean(0);
        GameAnalytics.setEnabledManualSessionHandling(flag);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureAvailableCustomDimensions01")) {
        String json = args.getString(0);
        ArrayList<String> list = new ArrayList<String>();

        try
        {
            JSONArray jsonArray = new JSONArray(json);
            for(int i = 0; i < jsonArray.length(); ++i)
            {
                list.add(jsonArray.getString(i));
            }
        }
        catch (JSONException e)
        {
        }

        GameAnalytics.configureAvailableCustomDimensions01(list.toArray(new String[0]));
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureAvailableCustomDimensions02")) {
        String json = args.getString(0);
        ArrayList<String> list = new ArrayList<String>();

        try
        {
            JSONArray jsonArray = new JSONArray(json);
            for(int i = 0; i < jsonArray.length(); ++i)
            {
                list.add(jsonArray.getString(i));
            }
        }
        catch (JSONException e)
        {
        }

        GameAnalytics.configureAvailableCustomDimensions02(list.toArray(new String[0]));
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("configureAvailableCustomDimensions03")) {
        String json = args.getString(0);
        ArrayList<String> list = new ArrayList<String>();

        try
        {
            JSONArray jsonArray = new JSONArray(json);
            for(int i = 0; i < jsonArray.length(); ++i)
            {
                list.add(jsonArray.getString(i));
            }
        }
        catch (JSONException e)
        {
        }

        GameAnalytics.configureAvailableCustomDimensions03(list.toArray(new String[0]));
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setCustomDimension01")) {
        String dimension = args.getString(0);
        GameAnalytics.setCustomDimension01(dimension);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setCustomDimension02")) {
        String dimension = args.getString(0);
        GameAnalytics.setCustomDimension02(dimension);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setCustomDimension03")) {
        String dimension = args.getString(0);
        GameAnalytics.setCustomDimension03(dimension);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setFacebookId")) {
        String id = args.getString(0);
        GameAnalytics.setFacebookId(id);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setGender")) {
        String gender = args.getString(0);
        GameAnalytics.setGender(gender);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("setBirthYear")) {
        int year = args.getInt(0);
        GameAnalytics.setBirthYear(year);
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("startSession")) {
        GameAnalytics.startSession();
        callback.success(); // Thread-safe.
        return true;
    } else if (action.equals("endSession")) {
        GameAnalytics.endSession();
        callback.success(); // Thread-safe.
        return true;
    }
    callback.error("Error"); // Thread-safe.
    return false;
  }
}

File: src/ios/GameAnalyticsPlugin.m
#import "GameAnalyticsPlugin.h"
#import "GameAnalytics.h"

@implementation GameAnalyticsPlugin

// interface implementation

- (void) initialize:(CDVInvokedUrlCommand*)command
{
    NSString *gameKey = [command.arguments objectAtIndex:0];
    NSString *gameSecret = [command.arguments objectAtIndex:1];
    [GameAnalytics initializeWithGameKey:gameKey gameSecret:gameSecret];
}
- (void) configureBuild:(CDVInvokedUrlCommand*)command
{
    NSString *build = [command.arguments objectAtIndex:0];
    [GameAnalytics configureBuild:build];
}
- (void) configureGameEngineVersion:(CDVInvokedUrlCommand*)command
{
    NSString *gameEngineVersion = [command.arguments objectAtIndex:0];
    [GameAnalytics configureEngineVersion:gameEngineVersion];
}
- (void) configureSdkGameEngineVersion:(CDVInvokedUrlCommand*)command
{
    NSString *sdkVersion = [command.arguments objectAtIndex:0];
    [GameAnalytics configureSdkVersion:sdkVersion];
}
- (void) configureUserId:(CDVInvokedUrlCommand*)command
{
    NSString *uId = [command.arguments objectAtIndex:0];
    [GameAnalytics configureUserId:uId];
}

- (void) addBusinessEvent:(CDVInvokedUrlCommand*)command
{
    NSString *currency = [command.arguments objectAtIndex:0];
    NSInteger amount = [command.arguments objectAtIndex:1];
    NSString *itemType = [command.arguments objectAtIndex:2];
    NSString *itemId = [command.arguments objectAtIndex:3];
    NSString *cartType = [command.arguments objectAtIndex:4];
    
    [GameAnalytics addBusinessEventWithCurrency:currency
                                         amount:amount
                                       itemType:itemType
                                         itemId:itemId
                                       cartType:cartType
                                        receipt:nil];
}
- (void) addBusinessEventWithReceipt:(CDVInvokedUrlCommand*)command
{
    NSString *currency = [command.arguments objectAtIndex:0];
    NSInteger amount = [command.arguments objectAtIndex:1];
    NSString *itemType = [command.arguments objectAtIndex:2];
    NSString *itemId = [command.arguments objectAtIndex:3];
    NSString *cartType = [command.arguments objectAtIndex:4];
    NSString *receipt = [command.arguments objectAtIndex:5];
    
    [GameAnalytics addBusinessEventWithCurrency:currency
                                         amount:amount
                                       itemType:itemType
                                         itemId:itemId
                                       cartType:cartType
                                        receipt:receipt];
}
- (void) addResourceEvent:(CDVInvokedUrlCommand*)command
{
    int flowType = [[command.arguments objectAtIndex:0] intValue];
    NSString *currency = [command.arguments objectAtIndex:1];
    double amount = [[command.arguments objectAtIndex:2] doubleValue];
    NSString *itemType = [command.arguments objectAtIndex:3];
    NSString *itemId = [command.arguments objectAtIndex:4];
    
    [GameAnalytics addResourceEventWithFlowType:(GAResourceFlowType)flowType
                                       currency:currency
                                         amount:@(amount)
                                       itemType:itemType
                                         itemId:itemId];
}
- (void) addProgressionEvent:(CDVInvokedUrlCommand*)command
{
    int progressionStatus = [[command.arguments objectAtIndex:0] intValue];
    NSString *progression01 = [command.arguments objectAtIndex:1];
    NSString *progression02 = [command.arguments objectAtIndex:2];
    NSString *progression03 = [command.arguments objectAtIndex:3];
    
    [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus
                                              progression01:progression01
                                              progression02:progression02
                                              progression03:progression03];
}
- (void) addProgressionEventWithScore:(CDVInvokedUrlCommand*)command
{
    int progressionStatus = [[command.arguments objectAtIndex:0] intValue];
    NSString *progression01 = [command.arguments objectAtIndex:1];
    NSString *progression02 = [command.arguments objectAtIndex:2];
    NSString *progression03 = [command.arguments objectAtIndex:3];
    int score = [[command.arguments objectAtIndex:4] intValue];
    
    [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus
                                              progression01:progression01
                                              progression02:progression02
                                              progression03:progression03
                                              score:score];
}
- (void) addDesignEvent:(CDVInvokedUrlCommand*)command
{
    NSString *eventId = [command.arguments objectAtIndex:0];
    [GameAnalytics addDesignEventWithEventId:eventId];
}
- (void) addDesignEventWithValue:(CDVInvokedUrlCommand*)command
{
    NSString *eventId = [command.arguments objectAtIndex:0];
    double value = [[command.arguments objectAtIndex:1] doubleValue];
    [GameAnalytics addDesignEventWithEventId:eventId value:@(value)];
}

- (void) addErrorEvent:(CDVInvokedUrlCommand*)command
{
    int severity = [[command.arguments objectAtIndex:0] intValue];
    NSString *message = [command.arguments objectAtIndex:1];
    [GameAnalytics addErrorEventWithSeverity:(GAErrorSeverity)severity message:message];
}
- (void) setEnabledInfoLog:(CDVInvokedUrlCommand*)command
{
    BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
    [GameAnalytics setEnabledInfoLog:flag];
}
- (void) setEnabledVerboseLog:(CDVInvokedUrlCommand*)command
{
    BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
    [GameAnalytics setEnabledVerboseLog:flag];
}
- (void) setEnabledManualSessionHandling:(CDVInvokedUrlCommand*)command
{
    BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
    [GameAnalytics setEnabledManualSessionHandling:flag];
}

- (void) configureAvailableCustomDimensions01:(CDVInvokedUrlCommand*)command
{
    NSString *json = [command.arguments objectAtIndex:0];
    NSArray *dimensions = [NSJSONSerialization JSONObjectWithData:[json dataUsingEncoding:NSUTF8StringEncoding]
                                                     options:kNilOptions
                                                       error:nil];
    [GameAnalytics configureAvailableCustomDimensions01:dimensions];
}
- (void) configureAvailableCustomDimensions02:(CDVInvokedUrlCommand*)command
{
    NSString *json = [command.arguments objectAtIndex:0];
    NSArray *dimensions = [NSJSONSerialization JSONObjectWithData:[json dataUsingEncoding:NSUTF8StringEncoding]
                                                     options:kNilOptions
                                                       error:nil];
    [GameAnalytics configureAvailableCustomDimensions02:dimensions];
}
- (void) configureAvailableCustomDimensions03:(CDVInvokedUrlCommand*)command
{
    NSString *json = [command.arguments objectAtIndex:0];
    NSArray *dimensions = [NSJSONSerialization JSONObjectWithData:[json dataUsingEncoding:NSUTF8StringEncoding]
                                                     options:kNilOptions
                                                       error:nil];
    [GameAnalytics configureAvailableCustomDimensions03:dimensions];
}
- (void) setCustomDimension01:(CDVInvokedUrlCommand*)command
{
    NSString *dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension01:dimension];
}
- (void) setCustomDimension02:(CDVInvokedUrlCommand*)command
{
    NSString *dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension02:dimension];
}
- (void) setCustomDimension03:(CDVInvokedUrlCommand*)command
{
    NSString *dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension03:dimension];
}

- (void) setFacebookId:(CDVInvokedUrlCommand*)command
{
    NSString *facebookId = [command.arguments objectAtIndex:0];
    [GameAnalytics setFacebookId:facebookId];
}
- (void) setGender:(CDVInvokedUrlCommand*)command
{
    NSString *gender = [command.arguments objectAtIndex:0];
    [GameAnalytics setGender:gender];
}
- (void) setBirthYear:(CDVInvokedUrlCommand*)command
{
    int year = [[command.arguments objectAtIndex:0] intValue];
    [GameAnalytics setBirthYear:year];
}

- (void) startSession:(CDVInvokedUrlCommand*)command
{
    [GameAnalytics startSession];
}
- (void) endSession:(CDVInvokedUrlCommand*)command
{
    [GameAnalytics endSession];
}

@end